### PR TITLE
Generalize UniformOverMesh

### PR DESCRIPTION
**Pull request by jschueller**
A colleague wanted to use the UniformOverMesh idea, but with a normal distribution, to simulate constrained defects.
I propose a new class "TruncatedOverMesh" to generalize that (I dont know if "Truncated" is right in that context):

```
threshold = 0.2
inVars = ["x1", "x2"]
formula = "sin(x1)*sin(x2)"
f = ot.SymbolicFunction(inVars, [formula])
levelSet = ot.LevelSet(f, ot.Less(), threshold)
xMin = [-5.0] * 2
xMax = [5.0] * 2
box = ot.Interval(xMin, xMax)
mesh = ot.LevelSetMesher([51] * 2).build(levelSet, box, False)
dist = ot.Normal([0]*2, [2]*2)
distribution_2d = otexp.TruncatedOverMesh(dist, mesh)
```

io

- The sampling algorithm chooses a simplex according to its weighted probability, then samples the truncated distribution on the simplex box until it lands in the simplex domain (could be ineficient when the dimension increases or if some elements are very narrow due to the mesh being irregular but still better than just trying to sample the original distribution until it lands in the whole mesh), for discussion if better ideas come up.

- Also it instanciates many distributions/domains/mesh on the fly, so its probably not super fast yet.

### CODE REVIEW

**Comment by regislebrun** [comment]
Nice idea, but IMO we should have a dedicated meeting to discuss the numerical aspects of such a class. In particular, there are alternatives to the proposed algorithms for the computation of the simplices weights and the sampling algorithm.

**Comment by jschueller** [comment]
Update also introduces the SimplicialCubature integration algorithm from Genz to integrate the pdf:
```
import openturns as ot
import openturns.experimental as otexp

v1 = [0.0, 0.0, 0.0]
v2 = [1.0, 1.0, 1.0]
v3 = [0.0, 1.0, 1.0]
v4 = [0.0, 0.0, 1.0]
# simplex
S = [v1, v2, v3, v4]
simplicies = [[0, 1, 2, 3]]
mesh = ot.Mesh(S, simplicies)
# function to integrate
f = ot.SymbolicFunction(['x1', 'x2', 'x3'], ['exp(x1 + x2 + x3)'])
# integral of f
algo = otexp.SimplicialCubature()
value = algo.integrate(f, mesh)[0]
```

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 1]
```diff
@@ -0,0 +1,852 @@
...
+
+    AES[K] = error;
+    VLS[K] = value;
+
+    VL += value;
+    AE += error;
+    NV += nodeNumber;
+  }
+
+  UnsignedInteger FL = 0;
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:199]: FL is used as a Bool. It should be a Bool.
```
Bool FL = false;
```

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+      VL += VLS[K];
+      AE += AES[K];
+      NV += nodeNumber;
+    }
+
+    NV += DFCOST;
+    flatSize += NEW - 1;
+
+    // Check for error termination.
+    FL = 0;
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:254]:
```
FL = false;
```

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+    for (UnsignedInteger i = 0; i < outputDimension; ++ i)
+      if (AE[i] > std::max(EA, ER*abs(VL[i])))
+      {
+        FL = 1;
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:258]:
```
FL = true;
```

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 4]
```diff
@@ -0,0 +1,852 @@
...
+    VL += value;
+    AE += error;
+    NV += nodeNumber;
+  }
+
+  UnsignedInteger FL = 0;
+  for (UnsignedInteger i = 0; i < outputDimension; ++ i)
+    if (AE[i] > std::max(EA, ER*abs(VL[i])))
+    {
+      FL = 1;
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:203]:
```
FL = true;
```

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 5]
```diff
@@ -0,0 +1,852 @@
...
+      break;
+    }
+
+  while ((FL > 0) && (NV + DFCOST + 4*nodeNumber <= MXFS))
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:207]:
```
while (FL && (NV + DFCOST + 4*nodeNumber <= MXFS))
```

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 6]
```diff
@@ -0,0 +1,852 @@
...
+       for (UnsignedInteger j = 0; j < outputDimension; ++ j)
+         if (AES(i, j) > maxAES)
+         {
+           ID = i;
+           maxAES = AES(i, j);
+         }
+
+    VL -= VLS[ID];
+    AE -= AES[ID];
+    UnsignedInteger NEW = computeNewSubregions(dimension, outputDimension, F, ID, flatSize, flatVertices);
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:222]: const

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 7]
```diff
@@ -0,0 +1,852 @@
...
+  }
+  return result;
+}
+
+void SimplicialCubature::computeRuleValueAndError(const UnsignedInteger dimension, const Sample & simplexVertices,
+                                                  const Scalar volume, const UnsignedInteger outputDimension,
+                                                  const Function & function, const Matrix & G, const Matrix & W,
+                                                  const Indices & evalBudget, Point & value, Point & error) const
+{
+  Scalar RTMN = 1e-1;
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:320]: const

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 8]
```diff
@@ -0,0 +1,852 @@
...
+  Scalar SMALL = SpecFunc::Precision;
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:321]: const

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 9]
```diff
@@ -0,0 +1,852 @@
...
+  Scalar ERRCOF = 8;
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:322]: const
Magical value?

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 10]
```diff
@@ -0,0 +1,852 @@
...
+
+  const UnsignedInteger WTS = W.getNbRows();
+  const UnsignedInteger RLS = W.getNbColumns();
+  Matrix RULE(outputDimension, RLS);
+
+  for (UnsignedInteger K = 0; K < WTS; ++ K)
+    if (evalBudget[K] > 0)
+    {
+      Point sym(computePermutationSums(dimension, simplexVertices, outputDimension, function, G.getColumn(K)*Point(1, 1.0)));
+      // TODO: better solution?
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:332]: You should probably unroll the computation to improve the performance. Something like:
```
for (UnsignedInteger i = 0; i < outputDimension; ++ i)
{
  const Scalar scaledSymI = volume * sym[i];
  for (UnsignedInteger j = 0; j < RLS; ++ j)
    RULE(i, j) += scaledSymI * W(K, j);
}

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 11]
```diff
@@ -0,0 +1,852 @@
...
+        symM(i, 0) = sym[i];
+      RULE = RULE + volume * symM * W.getRow(K);
+    }
+  // Scale integral values and compute the error estimates.
+  value = Point(outputDimension);
+  error = Point(outputDimension);
+  for (UnsignedInteger I = 0; I < outputDimension; ++ I)
+  {
+    value[I] = RULE(I, 0);
+    Scalar NMBS = std::abs(value[I]);
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:344]: const

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 12]
```diff
@@ -0,0 +1,852 @@
...
+    Scalar NMCP = 0.0;
+    Scalar RT = RTMN;
+    for (SignedInteger K = RLS-1-((RLS+1)%2); K >= 2; K -= 2)
+    {
+      Scalar NMRL = std::max(std::abs(RULE(I, K)), std::abs(RULE(I, K-1)));
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:349]: const

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 13]
```diff
@@ -0,0 +1,852 @@
...
+    else
+    {
+      Scalar DFSMX = 0;
+      for(UnsignedInteger L = 0; L < dimension+1; ++ L)
+      {
+        if ((L != IS) && (L != JS)) 
+        {
+          IT = std::min(std::min(L, IS), JS);
+          JT = std::max(std::max(L, IS), JS);
+          UnsignedInteger LT = IS + JS + L - IT - JT;
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:447]: const

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 14]
```diff
@@ -0,0 +1,852 @@
...
+          Scalar DFR = FRTHDF(IT, LT) + FRTHDF(LT, JT);
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:448]: const

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 15]
```diff
@@ -0,0 +1,852 @@
...
+          if (DFR >= DFSMX)
+          {
+            DFSMX = DFR;
+            LS = L;
+          }
+        }
+      }
+      Scalar DIFIL = FRTHDF(std::min(IS,LS), std::max(IS,LS));
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:456]: const

File: lib/src/Base/Algo/SimplicialCubature.cxx [hunk 16]
```diff
@@ -0,0 +1,852 @@
...
+      Scalar DIFLJ = FRTHDF(std::min(JS,LS), std::max(JS,LS));
```
Comment (regislebrun) [inline lib/src/Base/Algo/SimplicialCubature.cxx:457]: const

**Review by regislebrun — APPROVED** [review]
Nice work, only minor remarks.

Some of the methods of SimplicialCubature are not easy to read (it is more Fortran code & spirit), I did my best but I may have miss something.
A little comment would be nice after each closing brace of code bloc with more than 5 lines ;-)

File: lib/src/Uncertainty/Distribution/TruncatedOverMesh.cxx [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+/* Compute the numerical range of the distribution given the parameters values */
+void TruncatedOverMesh::computeRange()
+{
+  const Interval range(Interval(mesh_.getLowerBound(), mesh_.getUpperBound()).intersect(distribution_.getRange()));
+  setRange(range);
+}
+
+
+/** Build the sample of vertices for i-th simplex */
+Sample TruncatedOverMesh::getSimplexVertices(const UnsignedInteger index) const
```
[the remark this answers is not in the document]
Comment (jschueller) [reply lib/src/Uncertainty/Distribution/TruncatedOverMesh.cxx:106]: its only called in a loop in the ctor and once per getRealization

**Comment by jschueller** [comment]
> A little comment would be nice after each closing brace of code bloc with more than 5 lines ;-)

yes, most is inherited from the Fortran code which is poor in comments

**Pull request merged**